### PR TITLE
Implement more itertools function.

### PR DESCRIPTION
**Pull request by choleraehyq**
* added `groupby`, `izip_longest`, `permutations`, `product`, `combinations` and `combinations_with_replacement`.
* added corresponding tests.
* made `chain` to be an iterator.  [Docs](https://docs.python.org/2.7/library/itertools.html#itertools.chain).

Signed-off-by: Cholerae Hu <choleraehyq@gmail.com>

### CODE REVIEW

File: lib/itertools.py [hunk 1]
```diff
@@ -59,6 +74,29 @@ def dropwhile(predicate, iterable):
...
+class groupby(object):
+  # [k for k, g in groupby('AAAABBBCCDAABBB')] --> A B C D A B
+  # [list(g) for k, g in groupby('AAAABBBCCD')] --> AAAA BBB CC D
+  def __init__(self, iterable, key=None):
+    if key is None:
+      key = lambda x: x
+    self.keyfunc = key
+    self.it = iter(iterable)
+    self.tgtkey = self.currkey = self.currvalue = object()
+  def __iter__(self):
```
Comment (trotterdylan) [inline lib/itertools.py]: Please add a single blank line between method definitions

File: lib/itertools.py [hunk 2]
```diff
@@ -96,6 +134,58 @@ def izip(*iterables):
   while iterators:
     yield tuple(map(next, iterators))
 
+class ZipExhausted(Exception):
```
Comment (trotterdylan) [inline lib/itertools.py]: And two blank lines between top level definitions

File: lib/itertools_test.py [hunk 3]
```diff
@@ -103,6 +103,67 @@ def TestISlice():
     got = tuple(itertools.islice(*args))
     assert got == want, 'tuple(islice%s) == %s, want %s' % (args, got, want)
 
+def TestIZipLongest():
```
Comment (trotterdylan) [inline lib/itertools_test.py]: Two blank lines between top level definitions please

**Comment by choleraehyq** [comment]
@trotterdylan Done.

**Pull request merged**